codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -249,56 +400,83 @@ private void writeGoAwayFrame(Http2GoAwayFrame frame, ChannelPromise promise) {
+        }
+
         if (frame instanceof Http2DataFrame) {
             Http2DataFrame dataFrame = (Http2DataFrame) frame;
-            http2Handler.encoder().writeData(http2HandlerCtx, frame.streamId(), dataFrame.content().retain(),
-                                             dataFrame.padding(), dataFrame.isEndStream(), promise);
+            http2Handler.encoder().writeData(http2HandlerCtx, frame.stream().id(), dataFrame.content().retain(),
+                                             dataFrame.padding(), dataFrame.endStream(), promise);
         } else if (frame instanceof Http2HeadersFrame) {
             writeHeadersFrame((Http2HeadersFrame) frame, promise);
         } else if (frame instanceof Http2ResetFrame) {
             Http2ResetFrame rstFrame = (Http2ResetFrame) frame;
-            http2Handler.resetStream(http2HandlerCtx, frame.streamId(), rstFrame.errorCode(), promise);
+            http2Handler.encoder().writeRstStream(http2HandlerCtx, frame.stream().id(), rstFrame.errorCode(), promise);
         } else {
             throw new UnsupportedMessageTypeException(frame);
         }
     }
 
     private void writeHeadersFrame(Http2HeadersFrame headersFrame, ChannelPromise promise) {
-        int streamId = headersFrame.streamId();
-        if (!isStreamIdValid(streamId)) {
-            final Endpoint<Http2LocalFlowController> localEndpoint = http2Handler.connection().local();
-            streamId = localEndpoint.incrementAndGetNextStreamId();
-            try {
-                // Try to create a stream in OPEN state before writing headers, to catch errors on stream creation
-                // early on i.e. max concurrent streams limit reached, stream id exhaustion, etc.
-                localEndpoint.createStream(streamId, false);
-            } catch (Http2Exception e) {
-                promise.setFailure(e);
+        final int streamId;
+        if (isStreamIdValid(headersFrame.stream().id())) {
+            streamId = headersFrame.stream().id();
+        } else {
+            final Http2Stream2Impl stream = (Http2Stream2Impl) headersFrame.stream();
+            final Http2Connection connection = http2Handler.connection();
+            streamId = connection.local().incrementAndGetNextStreamId();
+            if (streamId < 0) {
+                promise.setFailure(new Http2NoMoreStreamIdsException());
                 return;
             }
-            ctx.fireUserEventTriggered(new Http2StreamActiveEvent(streamId, headersFrame));
+            numBufferedStreams++;
+            // Set the stream id before completing the promise, as any listener added by a user will be executed
+            // before the below listener, and so the stream identifier is accessible in a user's listener.
+            stream.id(streamId);
+            // Ensure that the listener gets executed before any listeners a user might have attached.
+            // TODO(buchgr): Once Http2Stream2 and Http2Stream are merged this is no longer necessary.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)